docs/02.API-REFERENCE.md

```diff
@@ -741,24 +741,40 @@ typedef bool (*jerry_backtrace_callback_t) (jerry_backtrace_frame_t *frame_p, vo
 
 **Summary**
 
-The type information of the native pointer.
-It includes the free callback that will be called when associated JavaScript object is garbage collected. It can be left NULL in case it is not needed.
-
-Typically, one would create a `static const jerry_object_native_info_t` for
-each distinct C type for which a pointer is used with
-`jerry_set_object_native_pointer ()` and `jerry_get_object_native_pointer ()`.
-This way, each `const jerry_object_native_info_t *` pointer address value itself
-uniquely identifies the C type of the native pointer.
-
-See [jerry_get_object_native_pointer](#jerry_get_object_native_pointer)
-for a best-practice code example.
+Type information for native pointers. Since each native pointer has a type information,
+multiple native pointers can be assigned to an object, and these can be updated or
+deleted independently.
+
+The type information has a free callback, which is called when the object is freed
+by the garbage collector. If the callback is NULL, the application is not notified
+about the destruction of the object.
+
+The buffer pointed by the native pointer can have a fixed number of jerry values,
+which refer to other values as long as the object is alive. These jerry values
+must be stored in the beginning of the buffer and their number is defined by the
+ `number_of_references` member. Before a buffer is attached to an object by
+[jerry_set_object_native_pointer](#jerry_set_object_native_pointer), the
+values must be initialized to undefined by
+[jerry_native_pointer_init_references](#jerry_native_pointer_init_references).
+When a buffer is no longer attached to an object, the
+[jerry_native_pointer_finalize_references](#jerry_native_pointer_finalize_references)
+must be called to release the values. A single buffer can be attached to any
+number of living objects. When a buffer is currently attached to at least
+one object, the references can be updated by
+[jerry_native_pointer_set_reference](#jerry_native_pointer_set_reference).
+However, if the buffer is no longer attached to an object, the finalize function
+must be called even if the buffer is reattached to another object later. In this
+case, calling the init function after the finalization is optional, because the
+finalize function also initializes all values to undefined.
 
 **Prototype**
 
 ```c
 typedef struct
 {
-  jerry_object_native_free_callback_t free_cb;
+  jerry_object_native_free_callback_t free_cb; /**< the free callback of the native pointer */
+  uint32_t number_of_references; /**< the number of value references stored in the beginning
```

Comment:
Was wondering if the "reference" is a good term for this. Maybe "slot" would be better? However I'm not sure. For me the "references" would indicate something related to refcounting and this functionality is slightly different.

Reply:
This is a mixed list of strong and weak references, hence I give the name. I don't like the slots. It could be weak references, if we would restrinct these to null.object values.